docs/proposals/kubelet-qos.md

```diff
@@ -0,0 +1,142 @@
+- Change the memory overcommit mode to 1, so that using excess memory starts the OOM killer.
+
+Setting OOM_SCORE_ADJ for a container
+- Refactor existing ApplyOomScoreAdj to util/oom.go
+- To set OOM_SCORE_ADJ of a container, we loop through all processes in the container, and set OOM_SCORE_ADJ
+- We keep looping until the list of processes in the container stabilizes
+
+Pod configuration
+- We’re focusing on Docker in this implementation (not RKT)
+- OOM scores
+  - Note that the OOM score of a process is 10 times the % of memory the process consumes, adjusted by OOM_SCORE_ADJ, barring exceptions (e.g. process is launched by root). Processes with higher OOM scores are killed.
+  - The base OOM score is between 0 and 1000, so if process A’s OOM_SCORE_ADJ - process B’s OOM_SCORE_ADJ is over a 1000, then process A will always be OOM killed before B.
+- Best-effort
+  - CPU shares: 2
+  - Set OOM_SCORE_ADJ: 1000
+  - So best-effort pods will have an OOM_SCORE of 1000
+- Guaranteed
+  - CPU shares: (request / CPU resource) * shares per CPU
+  - Set OOM_SCORE_ADJ: -999
+  - So guaranteed pods will have an OOM_SCORE of 0 or 1
+- Burstable
+  - CPU shares: (request / CPU resource) * shares per CPU
+  - If total memory request > 99% of available memory, OOM_SCORE_ADJ: 1000
+  - Otherwise, set OOM_SCORE_ADJ to 995 - 10 * (% of memory requested)
+    - This ensures that the OOM_SCORE of burstable pods is > 1
+    - So burstable pods will be killed if they conflict with guaranteed pods
+    - If a burstable pod uses less memory than requested, its OOM_SCORE <= 995
+    - So best-effort pods will be killed if they conflict with burstable pods using less than requested memory
+    - If a burstable pod uses more memory than requested, its OOM_SCORE > 995
+    - If a burstable pod that uses more memory than requested conflicts with a burstable pod using less memory than requested, the former will be killed
+    - If a burstable pod that uses more memory than requested conflicts with a best-effort pod, the behavior is undefined
+- Pod infrastructure container
+  - OOM_SCORE_ADJ: -999
+- Kubelet, Docker, Kube-Proxy
+  - OOM_SCORE_ADJ: -1000 (won’t be OOM killed)
+  - Hack, in the future, we should place all pods into a separate cgroup, and set a limit on the memory they can consume
+
+Ensuring resource guarantees
+- After being configured suitably, the kernel will handle resource guarantees as specified in QoS Functionality
+
+## Issues and Assumptions
+
+- The QoS proposal assumes that swap memory is disabled. If swap is enabled, then resource guarantees (for pods that specify resource requirements) will not hold. For example, suppose 2 top-tier pods have reached their memory limit. They can start allocating memory on swap space. Eventually, if there isn’t enough swap space, processes in the pods might get killed. TODO: ensure that swap space is disabled on our cluster setups scripts.

```

Comment:
Use "guaranteed" rather than "top-tier", to be consistent with the new terminology.